pyproject.toml

```diff
@@ -209,3 +209,20 @@ exclude= '''
 sdist-only = []
 git-only = [".*", "asv_benchmarks", "azure-pipelines.yml", "benchmarks", "build_tools", "maint_tools"]
 default-ignore = false
+
+[tool.spin]
+package = "sklearn"  # name of your package
+commands = [
+  "spin.cmds.meson.build",
+  "spin.cmds.meson.test",
+  "spin.cmds.pip.install",
+]
+# [tool.spin.commands]
+# "Build" = [
+#   "spin.cmds.meson.build",
+#   "spin.cmds.meson.test"
+# ]
+# "Environments" = [
+#   "spin.cmds.meson.ipython",
```

Comment:
on numpy, when working with spin, one cannot run python/ipython/pytest manually. So it's not that it's a shortcut, it's more like that this is the only entry point. I'm okay with that, but then we need these commands here too.

Reply:
If you build with meson into, say, `build-install`, you cannot run python/ipython/pytest without setting some paths. Spin does that for you, unless you are using an editable install (which you can also get with meson, via `pip install -e --no-build-isolation`).

I.e., it's not because of spin that you cannot access other entry-points, but because of the build/install mechanism used.

Reply:
Thanks for clarifying this. From the perspective of the users/contributors, it doesn't matter what the tool does. When users run a command which they expect to install the package, then it should behave as if it's installed.

So in this case, we might want to make sure `spin build` actually installs the package in the environment, at least for scikit-learn.

Reply:
Perhaps it's worth splitting the users and contributors. Users would install via pip or conda, typically, whereas developers want a working build that gets updated when they modify the code. The advantage of editable installs come especially when testing with other packages (i.e., you can run another project's test suite, and it will pick up your dev package). If you're happy to have an editable install (which retriggers builds and differs in some other ways from a regular package install), then you can set `build` to the `spin.cmds.pip.install` built-in command.

Reply:
Sorry I had local modifications that I did not push which now folded this discussion oh well ...

I think the best way forward on this would be to have `spin build-editable` exist and probably not `spin build` to avoid that people mix both, i.e. with spin you can choose which commands you make available. We would also need the Meson verbose editable install  so we kind of need a custom command I think.

Now I "just" need to find a good example of custom command to put some kind of reality behind the words :wink:

Reply:
Tell me what behavior you need, and I'll whip up a minimal example.

Reply:
Got it. What does "build in editable mode" mean in this context?

Reply:
See https://github.com/scientific-python/spin/pull/192 ; with that in place, we can use the command as-is, and just set `--verbose` as the default option.

Reply:
Examples of how NumPy overrides `spin` commands, as well as implement their own: https://github.com/numpy/numpy/blob/main/.spin/cmds.py

Reply:
> Got it. What does "build in editable mode" mean in this context?

Yeah not great wording, I think it is more "install in editable mode" indeed